agent/drive_handler.go

```diff
@@ -0,0 +1,136 @@
+// Copyright 2018-2019 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License"). You may
+// not use this file except in compliance with the License. A copy of the
+// License is located at
+//
+//	http://aws.amazon.com/apache2.0/
+//
+// or in the "license" file accompanying this file. This file is distributed
+// on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+// express or implied. See the License for the specific language governing
+// permissions and limitations under the License.
+
+package main
+
+import (
+	"io/ioutil"
+	"path/filepath"
+	"strconv"
+	"strings"
+)
+
+const (
+	blockPath = "/sys/block"
+
+	blockMajorMinor = "dev"
+	blockSize       = "size"
+)
+
+type drive struct {
+	Name       string
+	DriveID    string
+	MajorMinor string
+	Size       int
+	Type       string
+}
+
+type driveHandler struct {
+	drives map[string]drive
+}
+
+// DiscoverDrives will use lsblk to populate the drives field in
+// the drive handler.
+func (d *driveHandler) DiscoverDrives() error {
+	names, err := getListOfBlockDeviceNames()
+	if err != nil {
+		return err
+	}
+
+	drives := map[string]drive{}
+	for _, name := range names {
+		d, err := buildDrive(name)
+		if err != nil {
+			return err
+		}
+
+		drives[name] = d
+	}
+
+	for k, v := range drives {
+		if !isStubDrive(v) {
+			continue
+		}
+
+		id, err := getDriveID(v)
+		if err != nil {
+			return err
+		}
+
+		v.DriveID = id
+		drives[k] = v
+	}
+
+	d.drives = drives
+	return nil
+}
+
+func (d driveHandler) GetDrive(id string) (drive, bool) {
```

Comment:
a small nit about function ordering in this file: I generally like to see things flow from most-abstract to most-concrete so that you get deeper into implementation details the further you read the file.  Accordingly, I'd suggest the following ordering for this file:

* GetDrive
* DiscoverDrives
* getListOfBlockDeviceNames
* buildDrive
* isStubDrive
* getDriveID